boa/src/builtins/array/mod.rs

```diff
@@ -98,8 +99,93 @@ impl Array {
-        // Set Prototype
+        // Delegate to the appropriate constructor based on the number of arguments
+        match args.len() {
+            0 => Array::construct_array_empty(this, context),
+            1 => Array::construct_array_length(this, &args[0], context),
+            _ => Array::construct_array_values(this, args, context),
+        }
+    }
+
+    /// No argument constructor for `Array`.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array-constructor-array
+    fn construct_array_empty(this: &Value, context: &mut Context) -> Result<Value> {
+        let prototype = context.standard_objects().array_object().prototype();
+
+        Array::array_create(this, 0, Some(prototype), context)
+    }
+
+    /// By length constructor for `Array`.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array-len
+    fn construct_array_length(
+        this: &Value,
+        length: &Value,
+        context: &mut Context,
+    ) -> Result<Value> {
+        let prototype = context.standard_objects().array_object().prototype();
+        let array = Array::array_create(this, 0, Some(prototype), context)?;
+
+        if !length.is_number() {
+            array.set_field("0".to_string(), Value::from(length));
+            array.set_field("length".to_string(), Value::from(1));
+        } else {
+            if length.is_double() {
+                return context.throw_range_error("Invalid array length");
+            }
+            array.set_field(
+                "length".to_string(),
+                Value::from(length.to_u32(context).unwrap()),
+            );
+        }
+
+        Ok(array)
+    }
+
+    /// From items constructor for `Array`.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array-items
+    fn construct_array_values(
+        this: &Value,
+        items: &[Value],
+        context: &mut Context,
+    ) -> Result<Value> {
         let prototype = context.standard_objects().array_object().prototype();
+        let array = Array::array_create(this, items.len() as u32, Some(prototype), context)?;
+
+        for (k, item) in items.iter().enumerate() {
+            array.set_field(k.to_string(), item.clone());
+        }
+
+        Ok(array)
+    }
+
+    /// Utility for constructing `Array` objects.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-arraycreate
+    fn array_create(
+        this: &Value,
+        length: u32,
+        prototype: Option<GcObject>,
+        context: &mut Context,
+    ) -> Result<Value> {
+        let prototype = match prototype {
+            Some(prototype) => prototype,
+            None => context.standard_objects().array_object().prototype(),
```

Comment:
I'm not really sure about the correctness of this line, some help would be appreciated here. It's supposed to be step 4 from https://tc39.es/ecma262/#sec-arraycreate.

Reply:
The idea here is that if `prototype` is `None`, we should set it to `%Array.prototype%`. I'm not sure if this is the best way to retrieve it, maybe @HalidOdat can confirm.

Reply:
> I'm not sure if this is the best way to retrieve it, maybe @HalidOdat can confirm.

Yes. this is the best way to get hold of the `Array.prototype` object